### PR TITLE
client: calls to _ll_fh_exists() should hold client_lock

### PR DESCRIPTION
**Pull request by vshankar**
Credit to Brad Hubbard for grabbing the stack trace.

Fixes: https://tracker.ceph.com/issues/67565

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -16060,18 +16058,24 @@ int64_t Client::ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov,
       return retval;
     }
 
+    retval = 0;
+    std::unique_lock cl(client_lock);
```
Comment (dparmar18) [inline src/client/Client.cc:16062]: do we really need unique_lock?

File: src/client/Client.cc [hunk 2]
```diff
@@ -15842,6 +15842,10 @@ int Client::ll_read(Fh *fh, loff_t off, loff_t len, bufferlist *bl)
     return -CEPHFS_ENOTCONN;
   }
 
+  /* We can't return bytes written larger than INT_MAX, clamp len to that */
+  len = std::min(len, (loff_t)INT_MAX);
```
Comment (dparmar18) [inline src/client/Client.cc:15846]: Was moving this intentional? I guess it should be fine to keep it right before `int r = _read(fh, off, len, bl);`.?

File: src/client/Client.cc (hunk 1 above)
Comment (vshankar) [reply src/client/Client.cc:16062]: Using a unique_lock is the preferred way to achieve such a thing -- unlocking and locking it back again.

File: src/client/Client.cc (hunk 2 above)
Comment (vshankar) [reply src/client/Client.cc:15846]: Technically, yes, but this block doesn't require it to be under the lock.

File: src/client/Client.cc (hunk 1 above)
Comment (dparmar18) [reply src/client/Client.cc:16062]: ah, got it.

File: src/client/Client.cc (hunk 2 above)
Comment (vshankar) [reply src/client/Client.cc:15846]: I did think about that -- the problem is this dereferences `fh`.

Comment (dparmar18) [reply src/client/Client.cc:15846]: oh `tout(cct) << (uintptr_t)fh << std::endl;`, got it.

**Review by dparmar18 — APPROVED** [review]
LGTM

**Review by lxbsz — APPROVED** [review]

**Comment by badone** [comment]
Jenkins test make check

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67626](https://tracker.ceph.com/issues/67626).

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20240820042758-debug

**Pull request merged**